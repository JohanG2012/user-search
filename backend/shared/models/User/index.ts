import { Document, model, Schema } from 'mongoose';

interface IName extends Document {
  title: string;
  first: string;
  last: string;
}

interface IPicture extends Document {
  large: string;
  medium: string;
  thumbnail: string;
}

interface IUser extends Document {
  name: IName;
  picture: IPicture;
  permission: boolean;
}

const userSchema = new Schema(
  {
    name: {
      title: { type: String, required: true },
      first: { type: String, required: true },
      last: { type: String, required: true },
    },
    picture: {
      large: { type: String, required: true },
      medium: { type: String, required: true },
      thumbnail: { type: String, required: true },
    },
    permission: { type: Boolean },
  },
  {
    versionKey: false,
  },
);

export default model<IUser>('User', userSchema);
