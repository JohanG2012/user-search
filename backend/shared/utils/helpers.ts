export const stripObjProps = (inputArr: object[], stripArr: string[]): object[] => {
  return inputArr.map(obj => {
    const newObj = { ...obj };
    stripArr.forEach(str => {
      delete newObj[str];
    });
    return newObj;
  });
};
