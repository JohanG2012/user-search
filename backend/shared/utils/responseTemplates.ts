import { Context } from '@azure/functions';
import { HttpStatusCode } from '../enums/HttpStatusCode';
import { IHttpResponse } from '../interfaces/IHttpResponse';
import { IBody } from '../interfaces/IBody';

export const DATABASE_CONNECTION_ERROR = (): IHttpResponse => {
  const body: IBody = {
    status: 'error',
    message: 'Unable to reach database',
    errorCode: 2001,
  };

  const response: IHttpResponse = {
    status: HttpStatusCode.ServiceUnavailable,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const VALIDATION_ERROR = () => {
  // TODO: Implement, use code 2002
};

export const INTERNAL_ERROR = (e: { message: string }, context: Context): IHttpResponse => {
  context.log(e.message); // Could be set to print to log files instead
  const body: IBody = {
    status: 'error',
    message:
      'Oops! Something unknown happend. Please open an issues with intstructions on how to reproduce this response.',
    errorCode: 2003,
  };
  const response: IHttpResponse = {
    status: HttpStatusCode.serverError,
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const NOT_FOUND = (): IHttpResponse => {
  const body: IBody = {
    status: 'failed',
    message: 'Could not find requested resource.',
  };

  const response: IHttpResponse = {
    status: HttpStatusCode.NotFound,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const CREATED = (): IHttpResponse => {
  const body: IBody = {
    status: 'success',
    message: 'Successfully created',
  };

  const response: IHttpResponse = {
    status: HttpStatusCode.Created,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const OK = (data: any, next?: string): IHttpResponse => {
  const body: IBody = {
    status: 'success',
    data,
  };

  if (next) {
    body.next = next;
  }

  const response: IHttpResponse = {
    status: HttpStatusCode.Ok,
    headers: {
      'Content-Type': 'applicaiton/json',
    },
    body: JSON.stringify(body),
  };
  return response;
};

export const DELETED = (): IHttpResponse => ({
  status: HttpStatusCode.NoContent,
});

export const UPDATED = (): IHttpResponse => ({
  status: HttpStatusCode.NoContent,
});
