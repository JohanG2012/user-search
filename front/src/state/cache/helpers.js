export const generateSearchCache = (state, action) => {
  const newCache = {};
  newCache[action.payload.queryString] = {
    data: action.payload.data,
    next: action.payload.next,
  };
  const search = {
    ...state.search,
    ...newCache,
  };

  let entries = Object.entries(search);
  if (entries.length > 50) {
    entries = entries.slice(entries.length - 50);
  }

  const newSearch = {};

  entries.forEach(([key, value]) => {
    newSearch[key] = value;
  });

  return {
    search: {
      ...newSearch,
    },
  };
};

export default {
  generateSearchCache,
};
