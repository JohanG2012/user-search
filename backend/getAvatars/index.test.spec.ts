import * as mongoose from 'mongoose';
import { MongoMemoryServer } from 'mongodb-memory-server';
import httpFunction from './index';
import User from '../shared/models/User';
import { stripObjProps } from '../shared/utils/helpers';

let mongoServer;

beforeAll(async () => {
  mongoServer = new MongoMemoryServer();
  const mongoUri = await mongoServer.getConnectionString();
  await mongoose.connect(mongoUri, err => {
    if (err) {
      console.error(err);
    }
  });
  let i;
  for (i = 0; i < 20; i++) {
    const fakeUser = {
      name: {
        title: 'Mr.',
        first: 'Johan',
        last: 'Gudmundsson',
      },
      picture: {
        large: 'https://randomuser.me/api/portraits/men/36.jpg',
        medium: 'https://randomuser.me/api/portraits/med/men/36.jpg',
        thumbnail: 'https://randomuser.me/api/portraits/thumb/men/36.jpg',
      },
      permission: false,
    };
    await new User(fakeUser).save();
  }
});

afterAll(async () => {
  mongoose.disconnect();
  await mongoServer.stop();
});

describe('GET /api/v1/avatars', () => {
  let context: any = {};
  let req: any = {};
  beforeEach(() => {
    context = {
      res: {},
      done: jest.fn(),
      log: jest.fn(),
    };

    req = {
      params: {},
      query: {},
    };
  });
  describe('without querystrings', () => {
    beforeEach(async () => {
      await httpFunction(context, req);
    });

    it('should match snapshot', async () => {
      // strip _id to avoid snapshot collison
      const parsed = JSON.parse(context.res.body);
      const striped = stripObjProps(parsed.data, ['_id']);
      parsed.data = striped;
      context.res.body = JSON.stringify(parsed);

      expect(context).toMatchSnapshot();
      expect(context.done).toHaveBeenCalledTimes(1);
    });

    describe('few fallbacks incase bad snapshot is saved', () => {
      it('should respond with 200 status', () => {
        expect(context.res.status).toEqual(200);
      });

      it('should respond with a list of 10 avatars', () => {
        expect(JSON.parse(context.res.body).data.length).toBe(10);
      });
    });
  });

  describe('close down database connection', () => {
    beforeEach(async () => {
      mongoose.disconnect();
      mongoServer.stop();
      await httpFunction(context, req);
    });

    it('should respond with 500 error', async () => {
      expect(context.res.status).toBe(500);
    });

    it('should match snapshot', async () => {
      expect(context).toMatchSnapshot();
      expect(context.done).toHaveBeenCalledTimes(1);
    });
  });
});
