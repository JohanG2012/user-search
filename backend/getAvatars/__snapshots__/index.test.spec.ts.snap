// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /api/v1/avatars without querystrings should match snapshot 1`] = `
Object {
  "done": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "log": [MockFunction],
  "res": Object {
    "body": "{\\"status\\":\\"success\\",\\"data\\":[\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\",\\"https://randomuser.me/api/portraits/thumb/men/36.jpg\\"]}",
    "headers": Object {
      "Content-Type": "applicaiton/json",
    },
    "status": 200,
  },
}
`;
